src/common/hashtable/src/tail_array.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Allocator;
+
+use super::table0::Entry;
+use super::traits::Keyable;
+use crate::container::Container;
+use crate::container::StackContainer;
+
+const SIZE: usize = 8192;
+
+pub struct TailArray<K, V, A: Allocator> {
+    allocator: A,
+    pub(crate) datas: Vec<StackContainer<Entry<K, V>, SIZE, A>>,
+    pub(crate) num_items: usize,
+}
+
+impl<K, V, A> TailArray<K, V, A>
+where
+    K: Keyable,
+    A: Allocator + Clone,
+{
+    pub fn new(allocator: A) -> Self {
+        Self {
+            datas: vec![],
+            num_items: 0,
+            allocator,
+        }
+    }
+
+    pub fn insert(&mut self, key: K) -> &mut Entry<K, V> {
+        let pos = self.num_items % SIZE;
```

Comment:
tail array is ring buffer?

Reply:
No, it's a `Vector<  [Entrys; N]  >`. 

If we enable the tail array, the new keys will be pushed into this vector without expanding the hashtable.

Reply:
Is it only for the last block?